thunder/dynamo/benchmark_utils.py

```diff
@@ -370,83 +404,127 @@ def _check_tensor(x):
+    return True, log_str
+
+
+def check_metrics(
     folder_path,
     report,
     compile_fn1: Callable,
     compile_fn2: Callable,
     timer_fn: Callable,
-    timer_name: str,
-    rtol=0.5,
-    atol=0.0,
+    perf_rtol=0.5,
+    perf_atol=0.0,
+    memory_usage_rtol=0.5,
+    memory_usage_atol=0.0,
+    stream: TextIO = sys.stdout,
 ):
     """
-    Check the timing of graph report using two different compilation specifications with the provided timer configuration
+    Check the timing and memory usage(if using WallTimeWithMemoryUsage) of graph report using two different compilation specifications with the provided timer configuration
     and generate a benchmark script if the difference exceeds the threshold.
     """
+    folder_path = Path(folder_path)
     graph_name = report.graph_name
     if not is_report_using_cuda(report):
-        print(f"{graph_name} doesn't use CUDA, skip benchmark {timer_name}")
-        return
-    filename1 = f"{graph_name}_{compile_fn1.name}_{timer_name}.py"
-    filename2 = f"{graph_name}_{compile_fn2.name}_{timer_name}.py"
+        stream.write(f"{graph_name} doesn't use CUDA, skip benchmark {timer_fn.name}")
+        return None, None
+    filename1 = f"{graph_name}_{compile_fn1.name}_{timer_fn.name}.py"
+    filename2 = f"{graph_name}_{compile_fn2.name}_{timer_fn.name}.py"
 
     def try_and_log_benchmark(compile_fn, filename):
         try:
             return report.run_benchmark(compile_fn, timer_fn)
         except Exception as e:
-            msg = f"Benchmark {timer_name} on {graph_name} using {compile_fn.name} failed with exception {e}, benchmark script failed_{filename} is saved"
-            print(msg)
+            msg = f"Benchmark {timer_fn.name} on {graph_name} using {compile_fn.name} failed with exception {e}, benchmark script failed_{filename} is saved"
+            stream.write(msg)
+            failed_folder = folder_path / "failed"
             report.write_benchmark(
-                folder_path, compile_fn, timer_fn, file_name=f"failed_{filename1}", extra_comment_str=msg
+                failed_folder, compile_fn, timer_fn, file_name=f"failed_{filename1}", extra_comment_str=msg
             )
             return None
 
     measure1 = try_and_log_benchmark(compile_fn1, filename1)
     measure2 = try_and_log_benchmark(compile_fn2, filename2)
 
     if measure1 is None or measure2 is None:
-        return
+        return measure1, measure2
 
-    record = False
+    perf_record = False
+    memory_record = False
     log_strs = ""
     for m1, m2, name in zip(measure1, measure2, ("forward", "backward")):
-        if m1 is None:
-            assert m2 is None
-            continue
-        ret = check_threshold_log(
-            m1.median, m2.median, compile_fn1.name, compile_fn2.name, f"{graph_name} {name}", timer_name, rtol, atol
+        if timer_fn.name == "WallTimeWithMemoryUsage":
```

Comment:
This is OK for now, but in the future I don't think we want to rely on particular names to know whether to check memory usage or not.

One idea is to create a base class, like, `MaxAllocatedMemoryTimer`, that has an abstract method like `get_max_allocated()`, and then this check could become `if isinstance(timer, MaxAllocatedMemoryTimer)`. Then if someone wanted to implement a customer max allocated memory timer they would just inherit from this class. Similarly if we wanted to support reserve memory timers then there could be a `MaxReservedMemoryTimer` base class and timers could even inherit from both if they wanted to.